randovania/game_description/resources/resource_collection.py

```diff
@@ -0,0 +1,153 @@
+from __future__ import annotations
+
+import copy
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from randovania.game_description.resources.resource_database import ResourceDatabase
+    from randovania.game_description.resources.resource_info import ResourceGain, ResourceInfo
+
+
+class ResourceCollection:
+    __slots__ = ("resource_bitmask", "_resource_array", "_existing_resources", "add_self_as_requirement_to_resources",
+                 "_damage_reduction_cache")
+    resource_bitmask: int
+    _resource_array: list[int]
+    _existing_resources: dict[int, ResourceInfo]
+    add_self_as_requirement_to_resources: bool
+    _damage_reduction_cache: dict[int, float] | None
+
+    def __init__(self):
+        self.resource_bitmask = 0
+        self._resource_array = [0]
+        self._existing_resources = {}
+        self.add_self_as_requirement_to_resources = False
+        self._damage_reduction_cache = None
+
+    @classmethod
+    def with_database(cls, database: ResourceDatabase) -> ResourceCollection:
+        result = cls()
+        result._resource_array = [0] * len(database.resource_by_index)
+        return result
+
+    def _resize_array_to(self, size: int):
+        self._resource_array.extend(
+            0 for _ in range(len(self._resource_array), size + 1)
+        )
+
+    def __getitem__(self, item: ResourceInfo):
+        resource_index = item.resource_index
+        try:
+            return self._resource_array[resource_index]
+        except IndexError:
+            self._resize_array_to(resource_index)
+            return self._resource_array[resource_index]
+
+    def __str__(self):
+        return f"<ResourceCollection with {self.num_resources} resources>"
+
+    @property
+    def _comparison_tuple(self):
+        return list(self.as_resource_gain()), self.add_self_as_requirement_to_resources
```

Comment:
while we're doing type checking, it seems a bit odd for `comparison_tuple` to be a list

Reply:
ah i misread lol